src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1239 @@
+	 * FunctionCallInfoData, instead of using FunctionCall2Coll, is often
+	 * cheaper.
+	 */
+	FunctionCallInfoData compareValueFunctionCall;
+	FunctionCallInfoData compareIntervalFunctionCall;
+} ClauseWalkerContext;
+
+static void PrunableExpressions(Node *originalNode, ClauseWalkerContext *context);
+static bool PrunableExpressionsWalker(Node *originalNode, ClauseWalkerContext *context);
+static List * ShardArrayToList(ShardInterval **shardArray, int length);
+static int PerformValueCompare(FunctionCallInfoData *compareFunctionCall, Datum a,
+							   Datum b);
+static int PerformCompare(FunctionCallInfoData *compareFunctionCall);
+
+static List * PruneOne(DistTableCacheEntry *cacheEntry, ClauseWalkerContext *context,
+					   PruningInstance *prune);
+static List * PruneWithBoundaries(DistTableCacheEntry *cacheEntry,
+								  ClauseWalkerContext *context,
+								  PruningInstance *prune);
+static List * ExhaustivePrune(DistTableCacheEntry *cacheEntry,
+							  ClauseWalkerContext *context,
+							  PruningInstance *prune);
+static int UpperShardBoundary(Datum partitionColumnValue,
+							  ShardInterval **shardIntervalCache,
+							  int shardCount, FunctionCallInfoData *compareFunction,
+							  bool includeMin);
+static int LowerShardBoundary(Datum partitionColumnValue,
+							  ShardInterval **shardIntervalCache,
+							  int shardCount, FunctionCallInfoData *compareFunction,
+							  bool includeMax);
+
+
+/*
+ * PruneShards returns all shards from a distributed table that cannot be
+ * proven to be eliminated by whereClauseList.
+ *
+ * For reference tables, the function simply returns the single shard that the
+ * table has.
+ */
+List *
+PruneShards(Oid relationId, Index rangeTableId, List *whereClauseList)
+{
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+	char partitionMethod = cacheEntry->partitionMethod;
+	Var *partitionColumn = NULL;
+	ClauseWalkerContext context = { 0 };
+	ListCell *pruneCell;
+	List *prunedList = NIL;
+	bool foundRestriction = false;
+
+	/* always return empty result if WHERE clause is of the form: false (AND ..) */
+	if (ContainsFalseClause(whereClauseList))
+	{
+		return NIL;
+	}
+
+	/* short circuit for reference tables */
+	if (partitionMethod == DISTRIBUTE_BY_NONE)
+	{
+		return ShardArrayToList(cacheEntry->sortedShardIntervalArray,
+								cacheEntry->shardIntervalArrayLength);
+	}
+
+
+	partitionColumn = PartitionColumn(relationId, rangeTableId);
+
+	context.partitionMethod = partitionMethod;
+	context.partitionColumn = partitionColumn;
+	context.currentPruningInstance = palloc0(sizeof(PruningInstance));
+
+	if (cacheEntry->shardIntervalCompareFunction)
+	{
+		/* initiate function call info once (allows comparators to cache metadata) */
+		InitFunctionCallInfoData(context.compareIntervalFunctionCall,
+								 cacheEntry->shardIntervalCompareFunction,
+								 2, DEFAULT_COLLATION_OID, NULL, NULL);
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("frak1")));
```

Comment:
Err, this obviously should have a smarter error message - Or actually probably just an assert.